### ISSUE TITLE
OIDC, I cannot add extra claims from userinfo endpoint

### ISSUE THREAD

**Issue by rcladmin** [aspnet/Security#1449]
In ASPNET Core 2.0, extra claims from the userinfo endpoint are not added to User.Claims in a MVC Client application. This worked in Core 1.1, what am I doing wrong? All I am getting is sid,sub,idp and email , all the other claims (including my custom claims) from the userinfo endpoint is missing. I checked the access_token int jwt.io and all the claims are there.

```csharp
services.AddAuthentication(options =>
            {
                options.DefaultScheme = CookieAuthenticationDefaults.AuthenticationScheme;
                options.DefaultChallengeScheme = OpenIdConnectDefaults.AuthenticationScheme;
            })
            .AddCookie()
            .AddOpenIdConnect(o =>
            {
                o.SignInScheme = CookieAuthenticationDefaults.AuthenticationScheme; 
                o.RequireHttpsMetadata = false;
                o.Authority = "xxxxx";
                o.ClientId = "xxxx";
                o.ClientSecret = "xxxx";
                o.ResponseType = "code id_token";
                o.GetClaimsFromUserInfoEndpoint = true;
                o.SaveTokens = true;
                o.SecurityTokenValidator = new JwtSecurityTokenHandler
                {
                    InboundClaimTypeMap = new Dictionary<string, string>()
                };
                o.TokenValidationParameters.NameClaimType = "email";
            });
```

**Comment by Tratcher** [comment]
I wouldn't expect any differences. If you hook into Events.OnUserInformationReceived can you share the resulting json (see .User)?
https://github.com/aspnet/Security/blob/5abcfe7e3d78f777d45503b50de5cea17cf9aa5e/src/Microsoft.AspNetCore.Authentication.OpenIdConnect/Events/OpenIdConnectEvents.cs#L64

**Comment by rcladmin** [comment]
The .User in Events.OnUserInformationReceived does show the additional claims as 'Children' tokens. However, these additional claims do not seem to be added to the Identity. I end up with only the claims from the original id_token and not the access_token.

I realised that if I removed 'hybrid' flow and delete the property o.ResponseType = "code id_token";
an changed the flow to 'implicit' it works. I read this in a previous post and tried it. Any way, the code that works is shown below:

1. Use 'implicit' flow
2. Remove Response.Type property from OpenIdConnectOptions 

```csharp
services.AddAuthentication(options =>
            {
                options.DefaultScheme = CookieAuthenticationDefaults.AuthenticationScheme;
                options.DefaultChallengeScheme = OpenIdConnectDefaults.AuthenticationScheme;
            })
            .AddCookie()
            .AddOpenIdConnect(o =>
            {
                o.SignInScheme = CookieAuthenticationDefaults.AuthenticationScheme; 
                o.RequireHttpsMetadata = false;
                o.Authority = "xxxxx";
                o.ClientId = "xxxx";
                o.ClientSecret = "xxxx";
                o.GetClaimsFromUserInfoEndpoint = true;
                o.SaveTokens = true;
                o.SecurityTokenValidator = new JwtSecurityTokenHandler
                {
                    InboundClaimTypeMap = new Dictionary<string, string>()
                };
                o.TokenValidationParameters.NameClaimType = "email";
            });
```
Is it that additional claims from the userinfo endpoint are not added for 'hybrid' flows? I did not try other flows. What do I need to do to get the 'hybrid' flow to work? It worked in ASP.NET Core 1.x.

**Comment by yosbeleg89** [comment]
I have exactly the same problem with the same configuration of @rcladmin. I resolved setting `AlwaysIncludeUserClaimsInIdToken = true` in the client configuration on the IdentityServer4 side, so it is not being obtained from the userinfo endpoint.

**Comment by rcladmin** [comment]
These are my findings:

1. 'Implicit' flow (response type : id_token) gets all the claims in the id_token (there is no access_token here) and is added to the identity. However, this flow does not cause a request to userinfo endpoint. Therefore, AlwaysIncludeUserClaimsInIdToken = true has no effect in the implicit flow. This works.

2. Hybrid flow (response type: code id_token) gets claims from id_token and adds it to identity. It causes a request from the userinfo endpoint and get the extra claims in the access_token, but it does not add it to the Identity.

**Comment by Tratcher** [comment]
Ok, I know what's different here. See https://github.com/aspnet/Security/pull/1124#issuecomment-280079805. 2.0 no longer adds all possible information from the user-info endpoint, it was causing major cookie bloat leading to login issues. There is now a system called ClaimActions where you can select which elements you want to map from the user info doc to claims. See OpenIdConnectOptions.ClaimActions. Here are some examples:
https://github.com/aspnet/Security/blob/5abcfe7e3d78f777d45503b50de5cea17cf9aa5e/src/Microsoft.AspNetCore.Authentication.OpenIdConnect/OpenIdConnectOptions.cs#L61-L64
https://github.com/aspnet/Security/blob/5abcfe7e3d78f777d45503b50de5cea17cf9aa5e/samples/SocialSample/Startup.cs#L179

**Comment by rcladmin** [comment]
@Tratcher 

Hey, It works! Thanks for your help. For those having the same problem this is my code.

1. Use 'hybrid' flow
2. Add the claims you want to include with  `o.ClaimActions.MapJsonKey`

`````csharp
.AddOpenIdConnect(o =>
            {
                o.SignInScheme = CookieAuthenticationDefaults.AuthenticationScheme;
                o.RequireHttpsMetadata = false;
                o.Authority = "xxxx";
                o.ClientId = "xxxx";
                o.ClientSecret = "xxxx";
                o.SaveTokens = true;
                o.SecurityTokenValidator = new JwtSecurityTokenHandler
                {
                    InboundClaimTypeMap = new Dictionary<string, string>()
                };
                o.TokenValidationParameters.NameClaimType = "email";
                o.ResponseType = OpenIdConnectResponseType.CodeIdToken;
                o.GetClaimsFromUserInfoEndpoint = true;
                o.ClaimActions.MapJsonKey("Contacts", "Contacts");
            });
````

**Comment by balkarov** [comment]
@Tratcher
Hi everyone!

I have the same issue but I try to include array (role claim).
So when I try to 
`ClaimActions.MapJsonKey("role", "role")` 
I get the Exception:

> Cannot cast Newtonsoft.Json.Linq.JArray to Newtonsoft.Json.Linq.JToken

I tried to use MapCustomJson but I don't know how to return the array, threre is only string return type, not the List<string>
`ClaimActions.MapCustomJson("role", jObject => string.Join(',', ((JArray) jObject["role"]).ToObject<List<string>>()));`

**Comment by Tratcher** [comment]
The existing extensions are for simple one-to-one json values to claims. In your case role is an array. Do you want that array stored as a single claim? If so you can just call ToString on it.

If you want multiple claims then you can create your own ClaimAction that does that.
https://github.com/aspnet/Security/blob/5abcfe7e3d78f777d45503b50de5cea17cf9aa5e/src/Microsoft.AspNetCore.Authentication.OAuth/Claims/ClaimAction.cs#L40

**Comment by rcladmin** [comment]
@balkarov 

Get the roles as a string[] : `o.ClaimActions.MapCustomJson("role", jobj => jobj["role"].ToString());`

This will return something like this example: "role" : ["Admin", "User", "Reader", "Contributor"]

**Comment by yosbeleg89** [comment]
It worked for me too. Thanks @Tratcher

**Comment by balkarov** [comment]
@Tratcher thank you! It is work for me.
But why microsoft did not add "Role" claims by default in Asp Core ? I this "Role" is very popular claims.

```
public class JsonKeyArrayClaimAction: ClaimAction
{
    /// <summary>
    /// Creates a new JsonKeyArrayClaimAction.
    /// </summary>
    /// <param name="claimType">The value to use for Claim.Type when creating a Claim.</param>
    /// <param name="valueType">The value to use for Claim.ValueType when creating a Claim.</param>
    /// <param name="jsonKey">The top level key to look for in the json user data.</param>
    public JsonKeyArrayClaimAction(string claimType, string valueType, string jsonKey) : base(claimType, valueType)
    {
        JsonKey = jsonKey;
    }

    /// <summary>
    /// The top level key to look for in the json user data.
    /// </summary>
    public string JsonKey { get; }
    
    public override void Run(JObject userData, ClaimsIdentity identity, string issuer)
    {
        var values = userData?[JsonKey];
        if (!(values is JArray)) return;

        foreach (var value in values)
        {
            identity.AddClaim(new Claim(ClaimType, value.ToString(), ValueType, issuer));
        }
    }
}
```

And after 

`o.ClaimActions.Add(new JsonKeyArrayClaimAction("role", "role", "role"));`

And after I can use `User.IsInRole`

**Comment by rcladmin** [comment]
@balkarov 

I had the same question about Roles and Claims and this is what I found out:

1. Claims are preferred for authorization than the traditional Membership/Roles we used in the past
2. Microsoft has provided a new table in Identity called RoleClaims as a 'fix' for traditional users of asp.net identity roles. Its mainly for backward compatibility.
3. **Authorization Policy** is based mainly on a User Claim and is the preferred method to protect controllers. You decorate your controller with [Authorize(Policy="ContactsRead'] instead of the traditional User.IsInRole.
4. As a example, you may have users in a 'Marketing' role, but some user can only read data, some can delete and some can add data. This cannot be accomplished by roles but can be accomplished by user claims. Each user in a role can also have different claims. And you use these claims for the security. Its a better system.
5. The concept of roles-based authorization is non-existent is modern identitiy systems (facebook, twitter, google, miscosoft, etc). Claim based authorization seems to be the defacto system.

In summary, roles authorization has been an 'asp,net identity thing' that came with the old asp,net identity, claims based authorization has been more  widely accepted and is the recommended method for authorization. 

If you are interested to read more:

https://stormpath.com/blog/tutorial-policy-based-authorization-asp-net-core

**Comment by tstojecki** [comment]
I know this has been closed, but JsonKeyArrayClaimAction solved my problem too and I could see that added to Claims extension inside the middleware. 
//cc: @Tratcher

**Comment by Tratcher** [comment]
@tstojecki feel free to open a new issue for that and/or send a PR.

**Comment by tstojecki** [comment]
PR sent #1501

**Comment by sguryev** [comment]
`JsonKeyClaimAction` is in the Microsoft.AspNetCore.Authentication.OAuth v2.1.1 out of the box
Usage: `options.ClaimActions.Add(new JsonKeyClaimAction(JwtClaimTypes.Role, null, JwtClaimTypes.Role));`

I just can't understand why they pass `"role"` value to the `valueType` parameter. Like here: https://github.com/aspnet/Security/issues/1449#issuecomment-332767846
I expected something like "String".

Actually I have dig to the sources and this value is used for the `Claim` object creation.
`/// <param name="valueType">The claim value type. If this parameter is null, then <see cref="F:System.Security.Claims.ClaimValueTypes.String"></see> is used.</param>`

So it should be rather `null` than `"role"`
And `"role"` value passed to the `valueType` parameter obviously leads to the following:
![image](https://user-images.githubusercontent.com/6978458/44301390-6eea5f00-a31e-11e8-8b6b-88ca947b3caf.png)
